*: support compiling tidb to wasm

lucklove opened the pull request:
Signed-off-by: lucklove <gnu.crazier@gmail.com>

<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/community/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->
This PR make it possible to compile tidb to wasm and use tidb in browser.

### What is changed and how it works?
- add a term to interact with tidb core directly instead of mysql protol
- add wasm platform support for some low-level operation source file
- add a web interface to interact with term
- add polyfill for golang's missing fs operation in wasm

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->
 - Manual test

Code changes

Side effects

Related changes

Release note

lucklove commented:
This PR works now, but need a little work to do until I remove "WIP".
```
cd wasm
GOOS=js GOARCH=wasm go build -o main.css
python -m SimpleHTTPServer 8000
```

siddontang commented on util/mathutil/mathutil_js.go:
```diff
@@ -0,0 +1,102 @@
+const MaxInt = int(MaxUint >> 1) 
+const MinInt = -MaxInt - 1
+
+func MaxUint64(x, y uint64) uint64 {
```
please add comments for the export functions and vars

lucklove replied:
Ok, thanks for reminding

5kbpers commented on go.mod:
```diff
@@ -50,7 +50,7 @@ require (
 	github.com/prometheus/client_model v0.0.0-20180712105110-5c3871d89910
```
seems to need updating `goleveldb` to `7683883`

syrusakbary commented:
> GOOS=js GOARCH=wasm go build -o main.css

I just inspected https://play.tidb.io and was a bit weird seeing a 13Mb css file `main.css`.
When inspecting, it was actually a wasm file: https://play.tidb.io/main.css

Any reason on naming it `main.css` rather than `main.wasm` ?

syrusakbary commented on Makefile:
```diff
@@ -224,6 +225,14 @@ benchdb:
 
+wasm:
+	mkdir -p wasm-dist
+	GOOS=js GOARCH=wasm $(GOBUILD) -ldflags '$(LDFLAGS)' -o wasm-dist/main.css wasm/*.go
```
```suggestion
	GOOS=js GOARCH=wasm $(GOBUILD) -ldflags '$(LDFLAGS)' -o wasm-dist/main.wasm wasm/*.go
```

syrusakbary commented on Makefile:
```diff
@@ -224,6 +225,14 @@ benchdb:
+	cp "$(GOROOT)/misc/wasm/wasm_exec.js" wasm-dist/wasm_exec.js
+	cp wasm/favicon.ico wasm-dist/favicon.ico
+	cp wasm/jquery.console.js wasm-dist/jquery.console.js
+	sed 's/82837504/$(shell stat -f%z wasm-dist/main.css)/g' wasm/index.html.tpl > wasm-dist/index.html
```
```suggestion
	sed 's/82837504/$(shell stat -f%z wasm-dist/main.wasm)/g' wasm/index.html.tpl > wasm-dist/index.html
```

syrusakbary commented on .gitignore:
```diff
@@ -1,6 +1,8 @@
 /tidb-server/tidb-server
 /tidb-server/debug
+/wasm-dist
+/wasm/main.css
```
```suggestion
/wasm/main.wasm
```

syrusakbary commented on wasm/index.html.tpl:
```diff
@@ -0,0 +1,261 @@
+    function unimplemented2(_1, _2, callback) { unimplemented(callback); }
+    function bootstrapGo() {
+        const go = new Go();
+        fetch("main.css")
```
```suggestion
        fetch("main.wasm")
```

Deardrops commented on wasm/index.html.tpl:
```diff
@@ -0,0 +1,261 @@
+    <meta charset="utf-8" />
+    <title>TiDB Playground</title>
+    <link rel="icon" href="/favicon.ico" type="image/x-icon">
+    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
```
How about use local static file? For running TiDB wasm in offline environment.

Deardrops commented on wasm/index.html.tpl:
```diff
@@ -0,0 +1,261 @@
+
+    .outer {
+        width: 100%
+    } */
```
Please remove these commented css

lucklove commented:
@syrusakbary thanks for comment. 
We use .css instead of .wasm because many CDNs don't compress .wasm file automatically and unfortunately our CDN is one of them. In fact, the size of main.css is not 13MB, it's about 80MB, after compression, it's 13MB.

syrusakbary commented:
> We use .css instead of .wasm because many CDNs don't compress .wasm file automatically and unfortunately our CDN is one of them. In fact, the size of main.css is not 13MB, it's about 80MB, after compression, it's 13MB.

Oh no! Thanks for the info :)

lucklove commented:
@syrusakbary thanks for suggestion. I'll add a comment to makefile

lucklove commented:
known issue: golang1.13 can't produce large wasm correctly
https://github.com/golang/go/issues/34395

5kbpers commented:
This PR is ready to merge, PTAL, thx! @zz-jason @siddontang @ngaut

5kbpers commented on go.mod:
```diff
@@ -78,3 +80,5 @@ require (
 
 replace github.com/pingcap/check => github.com/tiancaiamao/check v0.0.0-20191119042138-8e73d07b629d
+
+replace github.com/coreos/go-systemd => github.com/5kbpers/go-systemd v0.0.0-20191209150347-994f05092cc6
```
Note: here we forked `go-systemd` for building WASM.

lucklove commented:
We have fixed TiDB's Wasm [compatibility problem](https://github.com/pingcap/tidb/pull/14183) and moved this feature to a standalone [repo](https://github.com/pingcap/tidb-wasm), closing this PR.
